qlib/Logger.qm

```diff
@@ -0,0 +1,2731 @@
+            const EVENT_OPEN = 1;
+            #! logging event
+            const EVENT_LOG = 2;
+            #! close event
+            const EVENT_CLOSE = 3;
+        }
+
+        private:internal {
+            #! The lock to to ensure proper serialization with concurrent access
+            RWLock m_lock();
+
+            #! an inactive appender won't accept any logging request
+            bool m_active = False;
+
+            #! filter chain
+            list<LoggerFilter> m_filters;
+
+            #! Appender name
+            string m_name;
+
+            #! async queue
+            *LoggerAppenderQueue m_queue;
+        }
+
+        #! Creates the object
+        /**
+            @param name the appender name
+        */
+        constructor(string name = "") {
+            m_name = name;
+        }
+
+        destructor() {
+            #close();  # TODO: cannot call as the target object may be already destroyed
+        }
+
+        #! Returns the appender name
+        public string getName() {
+            return m_name;
+        }
+
+        #! Sets the appender queue.
+        /**
+            Data are passed to target device synchronously
+            unless queue is assigned. In this case data are queued and
+            dedicated process will process it in asynchronously
+
+            @param queue
+        */
+        public setQueue(*LoggerAppenderQueue queue) {
+            if (m_active) {
+                throw "LOGGER-ERROR", "Appender is opened";
+            }
+            m_queue = queue;
+        }
+
+        #! Returns async queue or @ref nothing when events are processed synchronously
+        public *LoggerAppenderQueue getQueue() {
+            return m_queue;
+        }
+
+        #! Open logging resources
+        /**
+            The appender must be opened (active) to accept any logging events
+        */
+        public open() {
+            if (!m_active) {
+                pushEvent(EVENT_OPEN, NOTHING);
+                m_active = True;
+            }
+        }
+
+        #! Releases any resources allocated by the appender and closes it
+        /**
+        */
+        public close() {
+            if (m_active) {
+                pushEvent(EVENT_CLOSE, NOTHING);
+                m_active = False;
+            }
+        }
+
+        #! Returns the value of the \c active flag, i.e. @ref True if the appender is open and therefore active
+        public bool isActive() {
+            return m_active;
+        }
+
+        #! Adds a filter to the chain
+        /**
+            @param filter the new filter to add
+            @param top if @ref True then the filter is added at the start of the filter chain, if @ref False (the default),
+            the filter is added at the end of the chain
+            @throw "LOGGER-ERROR" thrown if the filter is already in the list
+        */
+        public addFilter(LoggerFilter filter, bool top = False) {
+            AutoWriteLock awl(m_lock);
```

Comment:
open, close, setQueue are not locked and thus not thread safe, but adding a filter is. Let's make it either all thread safe or not, otherwise there are race conditions.